iotdb-core/ainode/ainode/core/manager/inference_manager.py

```diff
@@ -135,49 +141,167 @@ def infer(self, full_data, window_interval=None, window_step=None, **_):
 
 
 class InferenceManager:
-    ACCELERATE_MODEL_ID = "sundial"
-    # DEFAULT_DEVICE = "cpu"
-    DEFAULT_DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
-    DEFAULT_POOL_SIZE = (
-        0  # TODO: Remove these parameter by sampling model inference consumption
-    )
+    ACCELERATE_MODEL_ID = ["sundial", "timer_xl"]
+    DEFAULT_DEVICE = "cpu"
+    # DEFAULT_DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")
     WAITING_INTERVAL_IN_MS = (
         AINodeDescriptor().get_config().get_ain_inference_batch_interval_in_ms()
     )  # How often to check for requests in the result queue
+    MODEL_MEM_USAGE_MAP = (
+        AINodeDescriptor().get_config().get_ain_inference_model_mem_usage_map()
+    )  # the memory usage of each model in bytes
+    INFERENCE_MEMORY_USAGE_RATIO = (
+        AINodeDescriptor().get_config().get_ain_inference_memory_usage_ratio()
+    )  # the device space allocated for inference
+    INFERENCE_WEIGHT_OVERHEAD_RATIO = (
+        AINodeDescriptor().get_config().get_ain_inference_weight_overhead_ratio()
+    )  # the weight overhead ratio for inference, used to estimate the pool size
 
     def __init__(self):
         self._model_manager = ModelManager()
         self._result_queue = mp.Queue()
         self._result_wrapper_map = {}
         self._result_wrapper_lock = threading.RLock()
+        self._pool_init_lock = (
+            threading.Lock()
+        )  # used for lazy initialization of inference request pools
         # structure: {model_id: [(InferenceRequestPool, request_queue), ...]}
         self._request_pool_map: Dict[str, List[(InferenceRequestPool, mp.Queue)]] = {}
         self._stop_event = mp.Event()
-        self._init_inference_request_pool()
         self._result_handler_thread = threading.Thread(
             target=self._handle_results, daemon=True
         )
         self._result_handler_thread.start()
+        self._model_mem_usage_map: Dict[str, int] = (
+            {}
+        )  # store model memory usage for each model
+        # self._preload_model_benchmarks()
+
+    def _preload_model_benchmarks(self):
+        if "cuda" in str(self.DEFAULT_DEVICE):
+            model_id = self.ACCELERATE_MODEL_ID
+            mem_usage = self._measure_model_memory(model_id)
+            self._model_mem_usage_map[model_id] = mem_usage
+            logger.info(
+                f"[Inference] Preloaded benchmark for {model_id}, mem_usage={mem_usage/1024**2:.2f} MB"
+            )
+        else:
+            logger.warning(
+                f"[Inference] Skipped preloading benchmarks for {self.DEFAULT_DEVICE}, only supports CUDA currently"
+            )
+
+    def _measure_model_memory(self, model_id: str) -> int:
```

Comment:
Consider create a new util file for extracting these `measure, estimate, evaluate` methods, to avoid the `inference_manager.py` being too complex to understand